### DISCUSSION TITLE
fatal error json.hpp file not found

### DISCUSSION THREAD

**Discussion by carul17** [nlohmann/json#3798] [Q&A]
I am trying to link the NLohmann json library to my json interpreter via cmake. 

I keep getting the error :
`fatal error: 'nlohmann/json.hpp' file not found
#include <nlohmann/json.hpp>`

From the file : 

```
#include <nlohmann/json.hpp>
#include <fstream>
#include "InterpretJson.h"
#include "game.h"
using namespace std;
using json = nlohmann::json;

InterpretJson(string path){
    this->path = path;
    ifstream f(path);
    json jData = json::parse(f);
    f.close();
    this->data = jData;
}

Game interpret(Game& game){}
```

 The CMakeLists.txt in the directory which contains src/interpretJson.cpp and include/interpretJson.h.:

```
find_package(nlohmann_json 3.2.0 REQUIRED)

add_library(interpreter
    src/interpretJson.cpp
)


target_include_directories(interpreter
    PUBLIC
        $<INSTALL_INTERFACE:include>
        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
    PRIVATE
        ${CMAKE_CURRENT_SOURCE_DIR}/src
        ${nlohmann_json_INCLUDE_DIR}
)

target_link_libraries(interpreter
    PRIVATE 
        ${nlohmann_json_LIBRARIES}
)

set_target_properties(interpreter
                    PROPERTIES
                    LINKER_LANGUAGE CXX
                    CXX_STANDARD 17
)

install(TARGETS interpreter
    ARCHIVE DESTINATION lib
)
```

How do I fix this?

**Comment by sreerajchundayil** [comment]
Do you have nlohmann folder and the hpp inside it?

**Reply by carul17** [reply]
Yeah I do have the nlohmann folder with the hpp installed on my system. So I had a smaller test2.cpp program just to check out the library. In that, this was my cmake file 
`cmake_minimum_required(VERSION 3.2)

project(Interpreter)

find_package(nlohmann_json 3.2.0 REQUIRED)
add_executable(interpret test.cpp)
target_link_libraries(interpret PRIVATE nlohmann_json::nlohmann_json)

set_target_properties(interpret PROPERTIES CXX_STANDARD 17)`

This compiled and ran perfectly however there were red squiggly lines below `#include <nlohmann/json.hpp>` which said `cannot open source file "nlohmann/json.hpp"`
However, like I said, it compiled and ran with all the json functionality working fine.

So for my main question, the problem may be related to the fact that I am trying to create a library instead of an executable. I am saying this because in the cmake which works, I am only creating an executable with just the cpp file. For my main question, I am trying to create a library with .cpp and .h files.

**Reply by gregmarr** [reply]
> This compiled and ran perfectly however there were red squiggly lines below #include <nlohmann/json.hpp> which said cannot open source file "nlohmann/json.hpp" However, like I said, it compiled and ran with all the json functionality working fine.

If this is Visual Studio, like it sounds, then this is just a difference between what the intellisense compiler recognizes and the regular compiler recognizes.  In a perfect world, they'd match perfectly, but unfortunately, this is not.
